File: item-list/src/main/java/com/gurunars/item_list/FetcherPermutations.java
package com.gurunars.item_list;

import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import java8.util.function.BiFunction;

class FetcherPermutations<ItemType extends Item> {

    @Inject
    private
    BiFunction<
            List<ItemType>,
            List<ItemType>,
            List<ItemType>> intersectionFetcher =
            new OrderedIntersectionFetcher<>();

    private
    FetcherUnidirectionalPermutations<ItemType> fetcherUnidirectionalPermutations =
            new FetcherUnidirectionalPermutations<>();

    private
    FetcherComplexPermutation<ItemType> fetcherComplexPermutation =
            new FetcherComplexPermutation<>();

    List<? extends Change<ItemType>> get(
            List<ItemType> sourceMiddle,
            List<ItemType> targetMiddle) {
        List<ItemType> intersectionSourceOrder =
                intersectionFetcher.apply(sourceMiddle, targetMiddle);
        List<ItemType> intersectionTargetOrder =
                intersectionFetcher.apply(targetMiddle, sourceMiddle);

        List<ChangeMove<ItemType>> moves = fetcherUnidirectionalPermutations.get(
                intersectionSourceOrder, intersectionTargetOrder, false);
        List<ChangeMove<ItemType>> movesReverse = fetcherUnidirectionalPermutations.get(
                intersectionSourceOrder, intersectionTargetOrder, true);

        if (moves == null && movesReverse == null) {

            return Collections.singletonList(
                    intersectionSourceOrder.isEmpty() || intersectionTargetOrder.isEmpty() ?
                            new ChangePersist<ItemType>() :
                            fetcherComplexPermutation.apply(
                                    intersectionSourceOrder,
                                    intersectionTargetOrder)
            );
        }

        if (moves == null) {
            return movesReverse;
        } else if (movesReverse == null) {
            return moves;
        }

        return movesReverse.size() < moves.size() ? movesReverse : moves;
    }
}

File: item-list/src/main/java/com/gurunars/item_list/FetcherPermutations.kt
package com.gurunars.item_list

import javax.inject.Inject

internal class FetcherPermutations<ItemType : Item> {

    @Inject
    private val intersectionFetcher = OrderedIntersectionFetcher<ItemType>()

    private val fetcherUnidirectionalPermutations = FetcherUnidirectionalPermutations<ItemType>()

    private val fetcherComplexPermutation = FetcherComplexPermutation<ItemType>()

    operator fun get(
            sourceMiddle: List<ItemType>,
            targetMiddle: List<ItemType>,
            startOffset: Int): List<Change<ItemType>> {
        val intersectionSourceOrder = intersectionFetcher.apply(sourceMiddle, targetMiddle)
        val intersectionTargetOrder = intersectionFetcher.apply(targetMiddle, sourceMiddle)

        val moves = fetcherUnidirectionalPermutations.get(
                intersectionSourceOrder, intersectionTargetOrder, startOffset, false)
        val movesReverse = fetcherUnidirectionalPermutations.get(
                intersectionSourceOrder, intersectionTargetOrder, startOffset, true)

        if (moves == null && movesReverse != null) {
            return movesReverse
        } else if (movesReverse == null && moves != null) {
            return moves
        } else if (moves != null && movesReverse != null) {
            return if (movesReverse.size < moves.size) movesReverse else moves
        } else {
            return listOf(if (intersectionSourceOrder.isEmpty() || intersectionTargetOrder.isEmpty())
                ChangePersist<ItemType>()
            else
                fetcherComplexPermutation.apply(
                        intersectionSourceOrder,
                        intersectionTargetOrder,
                        startOffset))
        }

    }
}
